### PR TITLE
Initial Pull Request

### LINKED ISSUE #113: Bring ansible-role-metasploit into cisagov and skeletonize it
# 💡 Summary #

Given that the Ansible role we use to install Metasploit ([GitHub](https://github.com/leesoh/ansible-role-metasploit), [Ansible Galaxy](https://galaxy.ansible.com/leesoh/metasploit)) isn't being maintained, we should bring that functionality into @cisagov as a properly skeletonized Ansible role.

## Motivation and context ##

We would then have a maintained Ansible role for installing Metasploit.

## Implementation notes ##

Please give credit to @leesoh for his original implementation.

### PR DESCRIPTION
**Pull request by king-alexander**
## 🗣 Description ##

These first commits modify `skeleton-ansible-role` to download and run the Metasploit framework installer.

## 💭 Motivation and context ##

This pull request was motivated by the discussion under https://github.com/cisagov/kali-packer/issues/112. While already resolved, it would be prudent of us to own our own Ansible role, in case there are any breaking changes in the future. 

- Together with cisagov/kali-packer#119, this PR resolves cisagov/kali-packer#113

## 🧪 Testing ##

I used a virtual environment (`pyenv 2.3.2`) running `python 3.9.7` and updated the default scenario ([84b1c38](https://github.com/cisagov/ansible-role-metasploit/pull/5/commits/84b1c380e51487521f6f74e0dbe77bf5502d887d)) to test my changes. Others can test this pull request with the command `molecule test --scenario-name default`. 

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

### CODE REVIEW

File: meta/main.yml
Comment (jsf9k) [inline meta/main.yml:10]: I would add a `metasploit` tag as well.

File: tasks/main.yml
Comment (jsf9k) [inline tasks/main.yml]: We prefer the fully-qualified collection names (FQCNs) to avoid warnings from the Ansible linter.  In this case the FQCN is `ansible.builtin.get_url`.

File: handlers/main.yml
Comment (jsf9k) [inline handlers/main.yml]: We prefer the fully-qualified collection names (FQCNs) to avoid warnings from the Ansible linter.  In this case the FQCN is `ansible.builtin.command`.

**Review by jsf9k — COMMENTED** [review]
I did a cursory pre-review and noted a few small things that you will want to change.

**Comment by jsf9k** [comment]
Have a look at the Molecule tests in [cisagov/ansible-role-burp-suite-pro](https://github.com/cisagov/ansible-role-burp-suite-pro).  I think that tool is installed outside of the package manager like this one is.

File: meta/main.yml
Comment (king-alexander) [reply meta/main.yml:10]: I amended [087d664](https://github.com/cisagov/ansible-role-metasploit/pull/5/commits/087d664f8a70cbe560f094b98ee47217b3118122) to include the tag.

File: tasks/main.yml
Comment (king-alexander) [reply tasks/main.yml]: I amended [087d664](https://github.com/cisagov/ansible-role-metasploit/pull/5/commits/087d664f8a70cbe560f094b98ee47217b3118122) to use the FQCN.

File: handlers/main.yml
Comment (king-alexander) [reply handlers/main.yml]: Same as https://github.com/cisagov/ansible-role-metasploit/pull/5#discussion_r927813243, I updated `handlers/main.yml` to use the FQCN.

File: meta/main.yml
Comment (jsf9k) [reply meta/main.yml:10]: The tags are no longer in alphabetical order.  Could you fix that?

File: handlers/main.yml
Comment (jsf9k) [inline handlers/main.yml]: The handler should delete the installer once it has been run.

File: molecule/default/tests/test_default.py
Comment (jsf9k) [inline molecule/default/tests/test_default.py]: The `msfinstall` process should create a directory somewhere for the MSF installation, perhaps under `/opt` or `/usr/local`.  If you run `which msfconsole` it might tell you where to look.

That would be a better test than simply checking to see if the installer is still around.

File: handlers/main.yml
Comment (jsf9k) [reply handlers/main.yml]: I would also move this code out of the handler and into the main tasks file.  Handlers are really intended to do things like restart services after modifying configs, whereas this installation is central to this Ansible role's raison d'être.

**Review by jsf9k — CHANGES_REQUESTED** [review]
I have a few requested changes, but the basic functionality is all there.  Nice work.

**Comment by jsf9k** [comment]
FWIW, you shouldn't have to run `molecule test --scenario-name default`.  `molecule` should use the `default` scenario by default, so you should only need to run `molecule test`.

File: meta/main.yml
Comment (jsf9k) [reply meta/main.yml:10]: Fixed via commit 9776036ee0fbbbc4913b24d0c129ebb82a2d112b.

File: molecule/default/tests/test_default.py
Comment (jsf9k) [inline molecule/default/tests/test_default.py]: We generally prefer long command line options in our code, since they are more descriptive and usually don't require a check of the `man` page to see what they do.
```suggestion
    assert host.run_expect([0], f'[ -n "$(ls --almost-all {dir_full_path})" ]')
```

Note that I have made the same fix in cisagov/ansible-role-burp-suite-pro#47.

File: meta/main.yml
Comment (jsf9k) [inline meta/main.yml]: You remove several platforms here, but the same platforms are still active in the `molecule/default/molecule.yml` file.  Do you intend to remove these platforms or not?

If the current code works for the other platforms then I suggest putting them back in.  We might as well support platforms that require no extra work on our part.

File: tasks/main.yml
Comment (jsf9k) [inline tasks/main.yml]: - `become: true` is the default, so you need not specify it.
- There is no harm in re-downloading the installer, and doing so avoids triggering the "handler" warning from `ansible-lint`.
```suggestion
- name: Download Metasploit Framework Installer
  ansible.builtin.get_url:
    url: https://raw.githubusercontent.com/rapid7/metasploit-omnibus/master/config/templates/metasploit-framework-wrappers/msfupdate.erb
    dest: /tmp/msfinstall
    mode: 0755

- name: Install Metasploit Framework
  ansible.builtin.command: /tmp/msfinstall
```

File: .ansible-lint
Comment (jsf9k) [inline .ansible-lint]: This addition is unnecessary if you accept my changes in `tasks/main.yml`.
```suggestion
```

**Review by jsf9k — CHANGES_REQUESTED** [review]
A few more recommendations...

File: molecule/default/tests/test_default.py
Comment (jsf9k) [reply molecule/default/tests/test_default.py]: Fixed in commit 238e0040d41ec51d85c64b408a4a34655fd5b4db.

File: handlers/main.yml
Comment (jsf9k) [reply handlers/main.yml]: Fixed in commit 9776036ee0fbbbc4913b24d0c129ebb82a2d112b.

File: molecule/default/tests/test_default.py
Comment (king-alexander) [reply molecule/default/tests/test_default.py]: Yes, I copied from https://github.com/cisagov/ansible-role-burp-suite-pro/blob/eb58b71740fdfeb7405e0fc6754d932ddfb67f2a/molecule/default/tests/test_default.py#L21

File: meta/main.yml
Comment (king-alexander) [reply meta/main.yml]: I did intend to remove those platforms since this role is supposed to primarily support Kali. But I agree! I'll add them back.

**Comment by king-alexander** [comment]
@jsf9k Now the linter throws `no-changed-when`. What would you recommend?

File: meta/main.yml
Comment (jsf9k) [reply meta/main.yml]: Fixed in commit f97c48e.

**Comment by jsf9k** [comment]
> @jsf9k Now the linter throws `no-changed-when`. What would you recommend?

Aargh.  Taking a look now.

**Comment by jsf9k** [comment]
> > @jsf9k Now the linter throws `no-changed-when`. What would you recommend?
> 
> Aargh. Taking a look now.

Have a look at commits e29a0f9 and 6aedf2d and let me know what you think.

**Comment by king-alexander** [comment]
> Have a look at commits [e29a0f9](https://github.com/cisagov/ansible-role-metasploit/commit/e29a0f9170d2d88e9c099b610899541ebbf2f3fb) and [6aedf2d](https://github.com/cisagov/ansible-role-metasploit/commit/6aedf2db16734f7844cb28bc755a0fb9c8f1b17e) and let me know what you think.

LGTM 👍

**Review by jsf9k — APPROVED** [review]
Nice work @king-alexander!

**Comment by king-alexander** [comment]
> Nice work @king-alexander!

Same to you @jsf9k!

**Review by dav3r — APPROVED** [review]
This looks great to me - nice job! 👍

**Comment by jsf9k** [comment]
@king-alexander - we'll wait for @mcdonnnj's review before merging.

**Comment by jsf9k** [comment]
See also cisagov/kali-packer#119 (in progress).

File: defaults/main.yml
Comment (mcdonnnj) [inline defaults/main.yml]:
```suggestion
# The directory where Metasploit Framework is to be installed
```

File: molecule/default/tests/test_default.py
Comment (mcdonnnj) [inline molecule/default/tests/test_default.py]:
```suggestion
    """Test that the Metasploit Framework installer was removed."""
```

File: molecule/default/tests/test_default.py
Comment (mcdonnnj) [inline molecule/default/tests/test_default.py]:
```suggestion
    """Test that the Metasploit Framework was installed."""
```

File: README.md
Comment (mcdonnnj) [inline README.md]:
```suggestion
The [Metasploit Framework](https://github.com/rapid7/metasploit-framework) is a
project of [Rapid7](https://www.rapid7.com/) in collaboration with the open
source community.
```

File: meta/main.yml
Comment (mcdonnnj) [inline meta/main.yml]: This wording mirrors a [similar role](https://github.com/cisagov/ansible-role-burp-suite-pro).

```suggestion
  description: Download and install the Metasploit Framework
```

File: tasks/main.yml
Comment (mcdonnnj) [inline tasks/main.yml]:
```suggestion
- name: Check if the Metasploit Framework is already installed
```

File: tasks/main.yml
Comment (mcdonnnj) [inline tasks/main.yml]:
```suggestion
- name: Install the Metasploit Framework
```

File: tasks/main.yml
Comment (mcdonnnj) [inline tasks/main.yml]:
```suggestion
    - name: Download the Metasploit Framework installer
```

File: tasks/main.yml
Comment (mcdonnnj) [inline tasks/main.yml]: Given that we run this shell script blindly I think we should pin to the latest commit for the file. I'm also open to pinning to the current tag if tying it to a commit seems unwieldy. The commit or tag could also be a variable for the role with a default so it's easy to override as needed. I am mostly just against blindly pulling down whatever has been merged into the `master` branch. Thoughts @cisagov/team-ois?

```suggestion
        url: https://raw.githubusercontent.com/rapid7/metasploit-omnibus/68a9de2b15dea8f3c7b1cf841874dee7ca4754f5/config/templates/metasploit-framework-wrappers/msfupdate.erb
```

File: tasks/main.yml
Comment (mcdonnnj) [inline tasks/main.yml]:
```suggestion
    - name: Install the Metasploit Framework
```

File: tasks/main.yml
Comment (mcdonnnj) [inline tasks/main.yml]:
```suggestion
    - name: Delete the Metasploit Framework installer
```

**Review by mcdonnnj — CHANGES_REQUESTED** [review]
This is a solid implementation. This is some strong work 💪💪💪. I mostly have some grammar/consistency suggestions but I did have a suggestion that I would like a collective decision around.

File: tasks/main.yml
Comment (dav3r) [reply tasks/main.yml]: I'm on board with pinning to a version and having a default.  I don't have a strong preference for pinning to a commit vs. a tag.

Comment (jsf9k) [reply tasks/main.yml]: We don't do this elsewhere, and we were advised by the former tech council that the assessors need the latest and greatest.  This is why this software is not being installed via the Kali Linux package in the first place.  To my knowledge this previous order has not been overridden or rescinded, and certainly the tech council has not provided the commit or tag that they want to use.

I vote for leaving this as it is until advised otherwise.

Comment (mcdonnnj) [reply tasks/main.yml]: This is not the version of the Metasploit Framework that is installed. This is the specific version of the installer *only*. The installer sets up the package manager and installs whatever is found there except for macOS which downloads the latest package available directly from the Metasploit website.

Comment (jsf9k) [reply tasks/main.yml]: OK, I'm fine with pinning and having a default.  We shouldn't do it via your commit, though, but via a role `default`.

Comment (jsf9k) [reply tasks/main.yml]: Please see commit 035f96c.

**Review by dav3r — APPROVED** [review]
👍 👍

File: defaults/main.yml
Comment (mcdonnnj) [inline defaults/main.yml]: Take-it-or-leave-it suggestion.

```suggestion
# Metasploit Framework installer to use.
```

File: meta/main.yml
Comment (mcdonnnj) [inline meta/main.yml:10]: Just a thought (I saw a role to install Nexpose that uses the tag).

```suggestion
    - rapid7
    - security
```

**Review by mcdonnnj — APPROVED** [review]
A couple minor suggestions. I would like to have an issue to get rid of the installer script in the future. The script simply adds a package repository and uses the system's package installer to install the `metasploit-framework` package. I'd rather have that behavior clearly defined and implemented in this role instead of retrieving and running an installer script.

File: defaults/main.yml
Comment (jsf9k) [reply defaults/main.yml]: As this isn't a complete sentence, I don't think a period is warranted.  That said, I won't have a conniption fit if one is added.

**Comment by jsf9k** [comment]
> I would like to have an issue to get rid of the installer script in the future. The script simply adds a package repository and uses the system's package installer to install the `metasploit-framework` package. I'd rather have that behavior clearly defined and implemented in this role instead of retrieving and running an installer script.

@king-alexander, so you want to create this issue or go ahead and just go ahead and make the change @mcdonnnj suggested?

**Comment by king-alexander** [comment]
> @king-alexander, so you want to create this issue or go ahead and just go ahead and make the change @mcdonnnj suggested?

I'll create the issue. To me, it sounds like a large enough unit of work to warrant one.

**Comment by jsf9k** [comment]
> > @king-alexander, so you want to create this issue or go ahead and just go ahead and make the change @mcdonnnj suggested?
> 
> I'll create the issue. To me, it sounds like a large enough unit of work to warrant one.

OK, let us know when you've created the issue (and referenced it via a comment at the top of `tasks.main.yml`) so we can put this PR to bed.

**Comment by king-alexander** [comment]
> OK, let us know when you've created the issue (and referenced it via a comment at the top of `tasks.main.yml`) so we can put this PR to bed.

Done!

File: defaults/main.yml
Comment (jsf9k) [reply defaults/main.yml]: @mcdonnnj, can we resolve this conversation?

Comment (mcdonnnj) [reply defaults/main.yml]: @jsf9k Sorry just looked at this again and this is just the second line of the full comment. I believe this should have a period on it to mirror the README per https://github.com/cisagov/ansible-role-metasploit/blob/021ed44ca8ab7b38e752535b9ee8594c4e9b7ace/README.md?plain=1#L18.

Comment (jsf9k) [reply defaults/main.yml]: OK, I committed the change.  Can you re-review?

File: defaults/main.yml
Comment (mcdonnnj) [inline defaults/main.yml]: I would prefer always quoting numeric-y values just to reinforce that they are strings in cases like this.

```suggestion
installer_version: "6.2.20"
```

**Review by mcdonnnj — APPROVED** [review]
I had one minor suggestion but otherwise LGTM ✔️

**Pull request merged**
